pingcap/tiflow
# sink (ticdc): add glue schema registry

### What problem does this PR solve?


Issue Number: ref #9467 

### What is changed and how it works?
**Why is this needed?**
Some users want to use AWS Glue Schema Registry as their schema registry when using Avro protocol.

**What is changed?**
This pull request implements an AWS Glue Schema Registry manager, which allows TiCDC to use the Glue Schema Registry when replicating messages using the Avro protocol.

**How does it works?**
A `glue-schema-registry-config` into the `sink.kafka-config`, you need to set it when you want to use glue schema registry.
To configure `glue-schema-registry-config` into `sink.kafka-config`, use the following example:
```toml
[sink]
[sink.kafka-config.glue-schema-registry-config]
region="us-west-1"  
registry-name="ticdc-test"
access-key="xxxx"
secret-access-key="xxxx"
tokne="xxxx"
```
`region` and `registry-name` are required, while `access-key`, `secret-access-key`, and `token` are optional. The best practice is to set your AWS credentials in environment variables or store them in the `~/.aws/credentials` file other than set them in the changefeed's config file. For more information, refer to the [AWS documentation](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-credentials).

### Check List 

#### Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

1. prepare changefeed config file
```toml
[sink]
[sink.kafka-config.glue-schema-registry-config]
region="us-west-1"
registry-name="ticdc-test"
access-key="xxxx"
secret-access-key="xxxx"
```
3. create a changefeed.
```
./cdc cli changefeed create --server=127.0.0.1:8300 --changefeed-id="kafka-avro" --sink-uri="kafka://xxxx" --config changefeed_glue.toml
```
4. Insert some data into upstream TiDB, do some alter table DDLs.
5. Run a kafka consumer and  use AWS official python deserializer to decode the message.
```
{'a': 7} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}]}
{'a': 8, 'name': 'dongmen'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
{'a': 9, 'name': 'dongmen'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
{'a': 10, 'name': 'guagua', 'email': 'xxx'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}, {"default": null, "name": "email", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
{'a': 12, 'name': 'drop email'} {"type": "record", "name": "t3", "namespace": "default.a", "fields": [{"name": "a", "type": {"type": "int", "connect.parameters": {"tidb_type": "INT"}}}, {"default": null, "name": "name", "type": ["null", {"type": "string", "connect.parameters": {"tidb_type": "TEXT"}}]}]}
```
6.  All data received and decoded are as expected.

#### Questions 

##### Will it cause performance regression or break compatibility?

##### Do you need to update user documentation, design documentation or monitoring documentation?

### Release note 

```release-note
support glue schema registry.
```
